### ISSUE TITLE
`to_vec` is applied recursively to entire structure, even when struct is only partially differentiable

### ISSUE THREAD

**Issue by gaurav-arya** [JuliaDiff/ChainRulesTestUtils.jl#256]
I am writing an `rrule` for `*(::Struct, arr)` where the structure and function look like,
```
struct Struct where {T<:Real}
     x # stuff that can't be differentiated
     y::T  
end

*(s::Struct, arr) = y * x(arr)
```
However, I am unable to test this `rrule`, even with a manually provided tangent for an instance of `Struct` that looks like `Tangent{Struct}(;x=NoTangent(), y=one(T))`.

The reason seems to be that finite differences tries to `to_vec` the instance of struct. Given that the struct is not completely ignored, only the field `x`, it ends up trying to `to_vec` the field `x` as well. But this field is a reference to a rather crazy mutable structure with circular references, and so I end up with an error, and am unable to test that the `rrule` is correct w.r.t. `y`.

(To make it more concrete, the structure in question is a `ScaledPlan` from `AbstractFFTs` and the field `x` refers to a primitive FFT plan from `FFTW`, which is mutable because of its `pinv` cache.)

Ideally, we shouldn't care at all what `x` is if it is marked as `NoTangent`, just as how the entire structure gets ignored if it is marked as `NoTangent`?

**Comment by devmotion** [comment]
> Tangent{Struct}(;x=NoTangent(), y=one(T))

Did you try `Tangent{typeof(mystruct)}(; y=randn(T))` or similar as well (ie without specifying `x`)?

**Comment by gaurav-arya** [comment]
I just did, and unfortunately that fails too. Looking at the code,
https://github.com/JuliaDiff/ChainRulesTestUtils.jl/blob/4050989ccdea6766f6546e1e2bcb2d77f36c6733/src/finite_difference_calls.jl#L23
it seems like the only influence of the tangent on the finite difference call is whether to ignore the full input or not?

**Comment by mzgubic** [comment]
>Ideally, it shouldn't matter what value is in the field x, since it is marked as NoTangent in the user-provided tangent? Just as how the entire input gets ignored if it is marked as NoTangent.

I agree this is the ideal behaviour. CRTU is not perfect, you can be almost certain that your rule is correct if tests pass, but tests failing could mean either an issue with the rule or an issue with CRTU.

Slightly less ideal solution, but quicker would be to try defining a `to_vec` method for the x field. (it's type piracy, but should be ok if just used in tests).

**Comment by gaurav-arya** [comment]
Any pointers on doing the necessary type piracy? 

For `test_frule`, I think I would need to overload `to_vec`, but what should the overload be so that the check will go through when the input tangent has a `NoTangent()` for field `x`? And for `test_rrule`, I think I might need to overload something else, like `ProjectTo`?

**Comment by mzgubic** [comment]
Sure, I would try with something like
```julia
function FiniteDifferences.to_vec(x::FFTPlan) # or whatever the type of field x is
        function FFTPlan_from_vec(x_vec::Vector)
            return x
        end
        return Bool[], FFTPlan_from_vec
    end
```
This should work for both the `frule` and `rrule` I believe

**Comment by gaurav-arya** [comment]
For posterity, here's what I needed to do to get `test_frule` and `test_rrule` to work on a structure with a non-differentiable, trouble nested field (in this case of type `InnerPlan`):
```
InnerPlan = Union{TestPlan, InverseTestPlan, TestRPlan, InverseTestRPlan}
function FiniteDifferences.to_vec(x::InnerPlan)
    function FFTPlan_from_vec(x_vec::Vector)
        return x
    end
    return Bool[], FFTPlan_from_vec
end
ChainRulesTestUtils.test_approx(::ChainRulesCore.AbstractZero, x::InnerPlan, msg=""; kwargs...) = true
ChainRulesTestUtils.rand_tangent(::AbstractRNG, x::InnerPlan) = ChainRulesCore.NoTangent()
```
Note that I needed the `test_approx` one as well, otherwise there was a call `zero(::InnerPlan)` being made in `test_rrule`. The `rand_tangent` one is not strictly necessary if one manually supplies the primal tangent, but is nicer as then `ChainRulesTestUtils` automatically fills out random tangents for all the other fields in the struct.

**Comment by oxinabox** [comment]
>  it seems like the only influence of the tangent on the finite difference call is whether to ignore the full input or not? (This code is called from test_frule, in which I also get an error)

That is correct for that.
It is also used for testing accumulation (`+`) some times.
Overloading `FiniteDifferences.to_vec` is the way to go.
Though we should probably make that one a little smarter, so this is needed less often.
